js/upgrade-db.js

```diff
@@ -0,0 +1,240 @@
+const Scalar = require("ffjavascript").Scalar;
+const commonjs = require("@hermeznetwork/commonjs");
+const commonjsOld = require("@hermeznetwork/commonjs-old");
+const SMT = require("circomlib").SMT;
+const poseidonHash = require("circomlib").poseidon;
+
+/**
+ * Upgrade rollupDb account to v1
+ */
+class UpgradeDb {
+
+    /**
+     * constructor
+     * @param {Object} rollupDb - RollupDb with old accounts
+     * @param {Number} nAccounts - maximum accounts to update in each proof
+     * @param {Number} nLevels - merkle tree depth
+     */
+    constructor(rollupDb, nAccounts, nLevels){
+        this.rollupDb = rollupDb;
+        this.nAccounts = nAccounts;
+        this.maxIdx = this.rollupDb.initialIdx;
+        this.initIdx = commonjs.Constants.firstIdx + 1;
+        this.isUpgraded = false;
+        this.lastBatch = this.rollupDb.lastBatch;
+        this.nLevels = nLevels;
+
+        // compute number of proofs to build
+        this.numProofs = Math.ceil((this.maxIdx - this.initIdx + 1) / this.nAccounts);
+        this.proofs = [];
+        this.newStateRoots = [];
+        this.finalIdxs = [];
+        this.imStateRoots = [];
+        this.imFinalIdxs = [];
+
+        this.dbState = new commonjs.SMTTmpDb(this.rollupDb.db);
+        this.stateTree = new SMT(this.dbState, this.rollupDb.stateRoot);
+    }
+
+    /**
+     * Perform upgrade for each account to new leaf state
+     */
+    async doUpgrade(){
+        let initKeyStateTree;
+        let finalKeystateTree;
+
+        for (let i = 0; i < this.numProofs; i++){
+            initKeyStateTree = this.initIdx + i*this.nAccounts;
+            finalKeystateTree = this.initIdx + (i+1)*this.nAccounts;
```

Comment:
maybe this should be `finalKeystateTree = this.initIdx + (i+1)*this.nAccounts -1 `? in order to upgrade `nAccounts` if not, the finalKeystateTree is the same as the `initKeyStateTree` of the next upgrade

`finalKeystateTree` --> `finalKeyStateTree`

Reply:
it depends in the loop inside the next function `_buildSingleProof`.
I'm iterating in `let i = initKey; i < finalKey; i++`, so max value is never reached and keys are not overlapped.

note: changed `finalKeystateTree` for `finalKeyStateTree`